File: Scalemon.FSM/ScaleStateMachine.vb
Imports System.Threading
Imports Microsoft.Extensions.Configuration
Imports Microsoft.Extensions.Logging
Imports Scalemon.Common

Imports Stateless

Public Class ScaleStateMachine

    Implements Scalemon.Common.IScaleStateMachine

    Private ReadOnly _fsm As StateMachine(Of Scalemon.Common.Enums.ScalesState, Scalemon.Common.Enums.Trigger)
    Private ReadOnly _weightReceivedTrigger As StateMachine(Of Scalemon.Common.Enums.ScalesState, Scalemon.Common.Enums.Trigger).TriggerWithParameters(Of Decimal)
    Private ReadOnly _onResetToZero As Func(Of Task)
    Private ReadOnly _onError As Func(Of Task)
    Private ReadOnly _logger As ILogger(Of ScaleStateMachine)
    Private ReadOnly _semaphore As New SemaphoreSlim(1, 1)
    Private ReadOnly _hystWeight As Decimal
    Private ReadOnly _minWeight As Decimal

    Private _zeroFlag As Boolean
    Private _errorFlag As Boolean
    Private _isInvalidWeight As Boolean
    Private _lastRaw As Decimal
    Private _semaphoreTime As Integer

    Public Sub New(logger As ILogger(Of ScaleStateMachine),
                       minWeight As Double,
                       hystWeight As Double,
                       semaphoreTimeMs As Integer,
                       onConnected As Func(Of Task),
                       onDisconnected As Func(Of Task),
                       onUnstable As Func(Of Task),
                       onResetToZero As Func(Of Task),
                       onZeroState As Func(Of Task),
                       onInvalidWeight As Func(Of Task),
                       onError As Func(Of Task),
                       onResetAlarm As Func(Of Task),
                       onRecord As Func(Of Decimal, Task))

        ' Читаем настройки
        _hystWeight = hystWeight
        _minWeight = minWeight
        _semaphoreTime = semaphoreTimeMs

        ' Создаём FSM
        _fsm = New StateMachine(Of Scalemon.Common.Enums.ScalesState, Scalemon.Common.Enums.Trigger)(Scalemon.Common.Enums.ScalesState.Disconnected)
        ' "Оборачиваем" обычный enum-триггер WeightReceived в параметризованный:
        _weightReceivedTrigger = _fsm.SetTriggerParameters(Of Decimal)(Scalemon.Common.Enums.Trigger.WeightReceived)
        _onResetToZero = onResetToZero
        _onError = onError
        _logger = logger

        ' 1) Подключение/отключение
        _fsm.Configure(Scalemon.Common.Enums.ScalesState.Disconnected) _
                .Permit(Scalemon.Common.Enums.Trigger.ScaleConnected, Scalemon.Common.Enums.ScalesState.Connected) _
                .Permit(Scalemon.Common.Enums.Trigger.DatabaseFailure, Scalemon.Common.Enums.ScalesState.DatabaseError)

        _fsm.Configure(Scalemon.Common.Enums.ScalesState.Connected) _
                .Permit(Scalemon.Common.Enums.Trigger.ScaleDisconnected, Scalemon.Common.Enums.ScalesState.Disconnected) _
                .Permit(Scalemon.Common.Enums.Trigger.ScaleAlarm, Scalemon.Common.Enums.ScalesState.ScaleError) _
                .Permit(Scalemon.Common.Enums.Trigger.DatabaseFailure, Scalemon.Common.Enums.ScalesState.DatabaseError) _
                .Permit(Scalemon.Common.Enums.Trigger.ScaleUnstable, Scalemon.Common.Enums.ScalesState.Unstable) _
                .PermitDynamic(Of Decimal)(_weightReceivedTrigger, AddressOf DetermineStateFromWeight) _
                .OnEntryAsync(Async Function()
                                  Await onConnected()
                              End Function) _
                .OnExitAsync(Async Function()
                                 Await onDisconnected()
                             End Function)

        ' 2) Нестабильное состояние
        _fsm.Configure(Scalemon.Common.Enums.ScalesState.Unstable) _
                .SubstateOf(Scalemon.Common.Enums.ScalesState.Connected) _
                .OnEntryFromAsync(Scalemon.Common.Enums.Trigger.ArduinoButtonPressed, Async Function()
                                                                                          If _errorFlag Then
                                                                                              _errorFlag = False
                                                                                              Await onResetAlarm()
                                                                                          End If
                                                                                      End Function) _
                .OnEntryAsync(Async Function()
                                  If _isInvalidWeight Then
                                      ' Сбрасываем сигнализацию, если была ошибка взвешивания
                                      _isInvalidWeight = False
                                      Await onResetAlarm()
                                  End If
                                  Await onUnstable()
                              End Function)

        ' 3) Стабилизированное состояние — суперкласс для весовых подкатегорий
        _fsm.Configure(Scalemon.Common.Enums.ScalesState.Stabilized) _
                .SubstateOf(Scalemon.Common.Enums.ScalesState.Connected) _
                .Permit(Scalemon.Common.Enums.Trigger.ScaleUnstable, Scalemon.Common.Enums.ScalesState.Unstable) _
                .Ignore(Scalemon.Common.Enums.Trigger.WeightReceived)

        ' 4) Категории внутри Stabilized
        _fsm.Configure(Scalemon.Common.Enums.ScalesState.NegativeWeight) _
                .SubstateOf(Scalemon.Common.Enums.ScalesState.Stabilized) _
                .OnEntryAsync(AddressOf HandleResetAttemptAsync)

        _fsm.Configure(Scalemon.Common.Enums.ScalesState.ZeroWeight) _
                .SubstateOf(Scalemon.Common.Enums.ScalesState.Stabilized) _
                .OnEntryAsync(Async Function()
                                  _zeroFlag = True
                                  Await onZeroState()
                              End Function)

        _fsm.Configure(Scalemon.Common.Enums.ScalesState.LightWeight) _
                .SubstateOf(Scalemon.Common.Enums.ScalesState.Stabilized) _
                .OnEntryAsync(AddressOf HandleResetAttemptAsync)

        _fsm.Configure(Scalemon.Common.Enums.ScalesState.InvalidWeight) _
                .SubstateOf(Scalemon.Common.Enums.ScalesState.Stabilized) _
                .OnEntryAsync(Async Function()
                                  _isInvalidWeight = True
                                  Await onInvalidWeight()
                              End Function)


        _fsm.Configure(Scalemon.Common.Enums.ScalesState.Recorded) _
                .SubstateOf(Scalemon.Common.Enums.ScalesState.Stabilized) _
                .OnEntryAsync(Async Function()
                                  _zeroFlag = False
                                  Await onRecord(_lastRaw)
                              End Function)

        _fsm.Configure(Scalemon.Common.Enums.ScalesState.ErrorAfterWeighing) _
                .SubstateOf(Scalemon.Common.Enums.ScalesState.Stabilized) _
                .Permit(Scalemon.Common.Enums.Trigger.ArduinoButtonPressed, Scalemon.Common.Enums.ScalesState.Unstable) _
                .OnEntryAsync(Async Function()
                                  _errorFlag = True
                                  Await onInvalidWeight()
                              End Function)

        ' 5) Аппаратная ошибка весов
        _fsm.Configure(Scalemon.Common.Enums.ScalesState.ScaleError) _
                .SubstateOf(Scalemon.Common.Enums.ScalesState.Connected) _
                .Permit(Scalemon.Common.Enums.Trigger.ScaleUnstable, Scalemon.Common.Enums.ScalesState.Unstable) _
                .Permit(Scalemon.Common.Enums.Trigger.ArduinoButtonPressed, Scalemon.Common.Enums.ScalesState.Unstable) _
                .PermitDynamic(Of Decimal)(_weightReceivedTrigger, AddressOf DetermineStateFromWeight) _
                .OnEntryAsync(Async Function()
                                  Await onError()
                              End Function)

        ' 6) Ошибка базы данных
        _fsm.Configure(Scalemon.Common.Enums.ScalesState.DatabaseError) _
                .OnEntryAsync(Async Function()
                                  Await onError() ' Зажигаем красную лампу
                              End Function) _
                .Permit(Scalemon.Common.Enums.Trigger.DatabaseRestored, Scalemon.Common.Enums.ScalesState.Unstable) ' Выход из ошибки при восстановлении БД
    End Sub

    Private Function DetermineStateFromWeight(raw As Decimal) As Scalemon.Common.Enums.ScalesState
        _lastRaw = raw
        If raw < 0D Then
            Return Scalemon.Common.Enums.ScalesState.NegativeWeight
        ElseIf raw = 0D Then
            Return Scalemon.Common.Enums.ScalesState.ZeroWeight
        ElseIf raw <= _hystWeight Then
            Return Scalemon.Common.Enums.ScalesState.LightWeight
        ElseIf raw <= _minWeight Then
            Return Scalemon.Common.Enums.ScalesState.InvalidWeight
        Else
            If _zeroFlag Then
                Return Scalemon.Common.Enums.ScalesState.Recorded
            Else
                Return Scalemon.Common.Enums.ScalesState.ErrorAfterWeighing
            End If
        End If
    End Function

    Public Async Function OnScaleConnectedAsync() As Task Implements IScaleStateMachine.OnScaleConnectedAsync
        If Not Await _semaphore.WaitAsync(2000) Then
            _logger.LogCritical("Не удалось захватить семафор FSM в течение 2 секунд. Автомат может быть заблокирован.")
            Return
        End If
        Try
            Await _fsm.FireAsync(Scalemon.Common.Enums.Trigger.ScaleConnected)
        Finally
            _semaphore.Release()
        End Try
    End Function

    Public Async Function OnScaleDisconnectedAsync() As Task Implements IScaleStateMachine.OnScaleDisconnectedAsync
        If Not Await _semaphore.WaitAsync(_semaphoreTime) Then
            _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.")
            Return
        End If
        Try
            Await _fsm.FireAsync(Scalemon.Common.Enums.Trigger.ScaleDisconnected)
        Finally
            _semaphore.Release()
        End Try
    End Function

    Private Async Function HandleResetAttemptAsync() As Task
        Dim resetException As Exception = Nothing
        Try
            ' Пытаемся выполнить асинхронную операцию
            Await _onResetToZero()
        Catch ex As Exception
            ' В блоке Catch только сохраняем исключение
            resetException = ex
        End Try
        ' Проверяем, была ли ошибка, уже ПОСЛЕ блока Catch
        If resetException IsNot Nothing Then
            _logger.LogError(resetException, "Автоматический сброс веса не удался.")
            Await _onError() ' Включаем красную лампу
            Await _fsm.FireAsync(Scalemon.Common.Enums.Trigger.ScaleAlarm) ' Переходим в состояние ошибки FSM
        End If
    End Function

    Public Async Function OnScaleUnstableAsync() As Task Implements IScaleStateMachine.OnScaleUnstableAsync
        If Not Await _semaphore.WaitAsync(_semaphoreTime) Then
            _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.")
            Return
        End If
        Try
            Await _fsm.FireAsync(Scalemon.Common.Enums.Trigger.ScaleUnstable)
        Finally
            _semaphore.Release()
        End Try
    End Function

    Public Async Function OnScaleAlarmAsync() As Task Implements IScaleStateMachine.OnScaleAlarmAsync
        If Not Await _semaphore.WaitAsync(_semaphoreTime) Then
            _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.")
            Return
        End If
        Try
            Await _fsm.FireAsync(Scalemon.Common.Enums.Trigger.ScaleAlarm)
        Finally
            _semaphore.Release()
        End Try
    End Function

    Public Async Function OnWeightReceivedAsync(raw As Decimal) As Task Implements IScaleStateMachine.OnWeightReceivedAsync
        If Not Await _semaphore.WaitAsync(_semaphoreTime) Then
            _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.")
            Return
        End If
        Try
            Await _fsm.FireAsync(_weightReceivedTrigger, raw)
        Finally
            _semaphore.Release()
        End Try
    End Function

    Public Async Function OnButtonPressedAsync() As Task Implements IScaleStateMachine.OnButtonPressedAsync
        If Not Await _semaphore.WaitAsync(_semaphoreTime) Then
            _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.")
            Return
        End If
        Try
            Await _fsm.FireAsync(Scalemon.Common.Enums.Trigger.ArduinoButtonPressed)
        Finally
            _semaphore.Release()
        End Try
    End Function

    Public Async Function OnDatabaseFailedAsync(ex As Exception) As Task Implements IScaleStateMachine.OnDatabaseFailedAsync
        If Not Await _semaphore.WaitAsync(_semaphoreTime) Then
            _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.")
            Return
        End If
        Try
            _logger.LogError(ex, "Получен сигнал о сбое в базе данных.")
            Await _fsm.FireAsync(Scalemon.Common.Enums.Trigger.DatabaseFailure)
        Finally
            _semaphore.Release()
        End Try
    End Function

    Public Async Function OnDatabaseRestoredAsync() As Task Implements IScaleStateMachine.OnDatabaseRestoredAsync
        If Not Await _semaphore.WaitAsync(_semaphoreTime) Then
            _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.")
            Return
        End If
        Try
            _logger.LogInformation("Получен сигнал о восстановлении базы данных.")
            Await _fsm.FireAsync(Scalemon.Common.Enums.Trigger.DatabaseRestored)
        Finally
            _semaphore.Release()
        End Try
    End Function
End Class

File: Scalemon.FSM/ScaleStateMachine.cs
using Microsoft.Extensions.Logging;
using Scalemon.Common;
using Stateless;
using System;
using System.Threading;
using System.Threading.Tasks;
using static Scalemon.Common.Enums;

namespace Scalemon.FSM
{


    public class ScaleStateMachine : IScaleStateMachine
    {

        private readonly StateMachine<Enums.ScalesState, Enums.Trigger> _fsm;
        private readonly StateMachine<Enums.ScalesState, Enums.Trigger>.TriggerWithParameters<decimal> _weightReceivedTrigger;
        private readonly Func<Task> _onResetToZero;
        private readonly Func<Task> _onError;
        private readonly ILogger<ScaleStateMachine> _logger;
        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
        private readonly decimal _hystWeight;
        private readonly decimal _minWeight;

        private bool _zeroFlag;
        private bool _errorFlag;
        private bool _isInvalidWeight;
        private decimal _lastRaw;
        private int _semaphoreTime;

        public ScaleStateMachine(ILogger<ScaleStateMachine> logger, double minWeight, double hystWeight, int semaphoreTimeMs, Func<Task> onConnected, Func<Task> onDisconnected, Func<Task> onUnstable, Func<Task> onResetToZero, Func<Task> onZeroState, Func<Task> onInvalidWeight, Func<Task> onError, Func<Task> onResetAlarm, Func<decimal, Task> onRecord)
        {

            // Читаем настройки
            _hystWeight = (decimal)hystWeight;
            _minWeight = (decimal)minWeight;
            _semaphoreTime = semaphoreTimeMs;

            // Создаём FSM
            _fsm = new StateMachine<Enums.ScalesState, Enums.Trigger>(Enums.ScalesState.Disconnected);
            // "Оборачиваем" обычный enum-триггер WeightReceived в параметризованный:
            _weightReceivedTrigger = _fsm.SetTriggerParameters<decimal>(Enums.Trigger.WeightReceived);
            _onResetToZero = onResetToZero;
            _onError = onError;
            _logger = logger;

            // 1) Подключение/отключение
            _fsm.Configure(Enums.ScalesState.Disconnected)
                .Ignore(Trigger.ScaleUnstable)
                .Ignore(Trigger.ScaleAlarm)
                .Ignore(Enums.Trigger.WeightReceived)
                .Permit(Enums.Trigger.ScaleConnected, Enums.ScalesState.Connected)
                .Permit(Enums.Trigger.DatabaseFailure, Enums.ScalesState.DatabaseError);


            _fsm.Configure(Enums.ScalesState.Connected)
                .Permit(Enums.Trigger.ScaleDisconnected, Enums.ScalesState.Disconnected)
                .Permit(Enums.Trigger.ScaleAlarm, Enums.ScalesState.ScaleError)
                .Permit(Enums.Trigger.DatabaseFailure, Enums.ScalesState.DatabaseError)
                .Permit(Enums.Trigger.ScaleUnstable, Enums.ScalesState.Unstable)
                .PermitDynamic(_weightReceivedTrigger, DetermineStateFromWeight)
                .OnEntryAsync(async () => await onConnected())
                .OnExitAsync(async () => await onDisconnected());

            // 2) Нестабильное состояние
            _fsm.Configure(Enums.ScalesState.Unstable)
                .SubstateOf(Enums.ScalesState.Connected)
                .OnEntryFromAsync(Enums.Trigger.ArduinoButtonPressed, async () => { if (_errorFlag) { _errorFlag = false; await onResetAlarm(); } }).OnEntryAsync(async () =>
                    {
                        if (_isInvalidWeight)
                            {
                                // Сбрасываем сигнализацию, если была ошибка взвешивания
                                _isInvalidWeight = false;
                                await onResetAlarm();
                            }
                    await onUnstable();
                    });

            // 3) Стабилизированное состояние — суперкласс для весовых подкатегорий
            _fsm.Configure(Enums.ScalesState.Stabilized).SubstateOf(Enums.ScalesState.Connected)
                .Permit(Enums.Trigger.ScaleUnstable, Enums.ScalesState.Unstable)
                .PermitDynamic(_weightReceivedTrigger, DetermineStateFromWeight);



            // 4) Категории внутри Stabilized
            _fsm.Configure(Enums.ScalesState.NegativeWeight)
                .SubstateOf(Enums.ScalesState.Stabilized)
                .OnEntryAsync(HandleResetAttemptAsync);


            _fsm.Configure(Enums.ScalesState.ZeroWeight)
                .SubstateOf(Enums.ScalesState.Stabilized)
                .OnEntryAsync(async () =>

                                {
                                    _zeroFlag = true;
                                    await onZeroState();
                                });

            _fsm.Configure(Enums.ScalesState.LightWeight)
                .SubstateOf(Enums.ScalesState.Stabilized)
                .OnEntryAsync(HandleResetAttemptAsync);

            _fsm.Configure(Enums.ScalesState.InvalidWeight)
                .SubstateOf(Enums.ScalesState.Stabilized)
                .OnEntryAsync(async () =>

                                {
                                    _isInvalidWeight = true;
                                    await onInvalidWeight();
                                });


            _fsm.Configure(Enums.ScalesState.Recorded)
                .SubstateOf(Enums.ScalesState.Stabilized)
                .OnEntryAsync(async () =>
                                {
                                    _zeroFlag = false;
                                    await onRecord(_lastRaw);
                                });

            _fsm.Configure(Enums.ScalesState.ErrorAfterWeighing)
                .SubstateOf(Enums.ScalesState.Stabilized)
                .Permit(Enums.Trigger.ArduinoButtonPressed, Enums.ScalesState.Unstable)
                .OnEntryAsync(async () =>
                                {
                                    _errorFlag = true;
                                    await onInvalidWeight();
                                });

            // 5) Аппаратная ошибка весов
            _fsm.Configure(Enums.ScalesState.ScaleError)
                .SubstateOf(Enums.ScalesState.Connected)
                .Permit(Enums.Trigger.ScaleUnstable, Enums.ScalesState.Unstable)
                .Permit(Enums.Trigger.ArduinoButtonPressed, Enums.ScalesState.Unstable)
                .PermitDynamic(_weightReceivedTrigger, DetermineStateFromWeight)
                .OnEntryAsync(async () => await onError());

            // 6) Ошибка базы данных
            _fsm.Configure(Enums.ScalesState.DatabaseError)
                .OnEntryAsync(async () => await onError())
                .Permit(Enums.Trigger.DatabaseRestored, Enums.ScalesState.Unstable);
        }

        private Enums.ScalesState DetermineStateFromWeight(decimal raw)
        {
            _lastRaw = raw;
            if (raw < 0m)
            {
                return Enums.ScalesState.NegativeWeight;
            }
            else if (raw == 0m)
            {
                return Enums.ScalesState.ZeroWeight;
            }
            else if (raw <= _hystWeight)
            {
                return Enums.ScalesState.LightWeight;
            }
            else if (raw <= _minWeight)
            {
                return Enums.ScalesState.InvalidWeight;
            }
            else if (_zeroFlag)
            {
                return Enums.ScalesState.Recorded;
            }
            else
            {
                return Enums.ScalesState.ErrorAfterWeighing;
            }
        }

        public async Task OnScaleConnectedAsync()
        {
            if (!await _semaphore.WaitAsync(2000))
            {
                _logger.LogCritical("Не удалось захватить семафор FSM в течение 2 секунд. Автомат может быть заблокирован.");
                return;
            }
            try
            {
                await _fsm.FireAsync(Enums.Trigger.ScaleConnected);
            }
            finally
            {
                _semaphore.Release();
            }
        }

        public async Task OnScaleDisconnectedAsync()
        {
            if (!await _semaphore.WaitAsync(_semaphoreTime))
            {
                _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.");
                return;
            }
            try
            {
                await _fsm.FireAsync(Enums.Trigger.ScaleDisconnected);
            }
            finally
            {
                _semaphore.Release();
            }
        }

        private async Task HandleResetAttemptAsync()
        {
            Exception resetException = null;
            try
            {
                // Пытаемся выполнить асинхронную операцию
                await _onResetToZero();
            }
            catch (Exception ex)
            {
                // В блоке Catch только сохраняем исключение
                resetException = ex;
            }
            // Проверяем, была ли ошибка, уже ПОСЛЕ блока Catch
            if (resetException is not null)
            {
                _logger.LogError(resetException, "Автоматический сброс веса не удался.");
                await _onError(); // Включаем красную лампу
                await _fsm.FireAsync(Enums.Trigger.ScaleAlarm); // Переходим в состояние ошибки FSM
            }
        }

        public async Task OnScaleUnstableAsync()
        {
            if (!await _semaphore.WaitAsync(_semaphoreTime))
            {
                _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.");
                return;
            }
            try
            {
                await _fsm.FireAsync(Enums.Trigger.ScaleUnstable);
            }
            finally
            {
                _semaphore.Release();
            }
        }

        public async Task OnScaleAlarmAsync()
        {
            if (!await _semaphore.WaitAsync(_semaphoreTime))
            {
                _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.");
                return;
            }
            try
            {
                await _fsm.FireAsync(Enums.Trigger.ScaleAlarm);
            }
            finally
            {
                _semaphore.Release();
            }
        }

        public async Task OnWeightReceivedAsync(decimal raw)
        {
            if (!await _semaphore.WaitAsync(_semaphoreTime))
            {
                _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.");
                return;
            }
            try
            {
                await _fsm.FireAsync(_weightReceivedTrigger, raw);
            }
            finally
            {
                _semaphore.Release();
            }
        }

        public async Task OnButtonPressedAsync()
        {
            if (!await _semaphore.WaitAsync(_semaphoreTime))
            {
                _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.");
                return;
            }
            try
            {
                await _fsm.FireAsync(Enums.Trigger.ArduinoButtonPressed);
            }
            finally
            {
                _semaphore.Release();
            }
        }

        public async Task OnDatabaseFailedAsync(Exception ex)
        {
            if (!await _semaphore.WaitAsync(_semaphoreTime))
            {
                _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.");
                return;
            }
            try
            {
                _logger.LogError(ex, "Получен сигнал о сбое в базе данных.");
                await _fsm.FireAsync(Enums.Trigger.DatabaseFailure);
            }
            finally
            {
                _semaphore.Release();
            }
        }

        public async Task OnDatabaseRestoredAsync()
        {
            if (!await _semaphore.WaitAsync(_semaphoreTime))
            {
                _logger.LogCritical("Не удалось захватить семафор FSM в течение _semaphoreTime секунд. Автомат может быть заблокирован.");
                return;
            }
            try
            {
                _logger.LogInformation("Получен сигнал о восстановлении базы данных.");
                await _fsm.FireAsync(Enums.Trigger.DatabaseRestored);
            }
            finally
            {
                _semaphore.Release();
            }
        }
    }
}